### PR TITLE
Add projected Hamiltonian utilities

### LINKED ISSUE #61: Add P'HP Hamiltonian projection utilities
Parent epic: #55

Planned PR branch/name: `pr-6-projected-hamiltonian-utils`

## Scope
- Add internal helpers `project_hamiltonian(H, projections; cutoff)` and `project_state(psi, projections; cutoff)`.
- Implement CoTenN-inspired DMRG preparation: build projections `P`, compute `H_eff = P' * H * P`, and project initial `psi`.
- Keep this internal and not yet wired into public `minimize`.

## Acceptance Criteria
- [ ] Dense small-system tests verify `P'HP` equals brute-force Hamiltonian masking.
- [ ] Projected state has zero amplitude on infeasible assignments for small systems.
- [ ] Prime-level cleanup is correct for ITensorMPS DMRG input.
- [ ] Existing unconstrained tests remain unchanged.

## Depends On
- #58 (Implement SumConstraint projection MPO)
- #59 (Implement not-equals and pairwise relation projection MPOs)
- #60 (Implement exactly-one projection MPO)

## Notes
Each issue is intended to be completed by one focused PR.
## Attribution
This work is inspired by the projection-MPO constraint framework in Sharma, Ritvik, Cheng Peng, Siddharth Dangwal, and Sara Achour. "CoTenN: Constrained Optimization with Tensor Networks." Proceedings of the ACM on Programming Languages 10, no. PLDI (2026): 629-655.

### PR DESCRIPTION
**Pull request by bernalde**
## Summary

- Add internal `project_hamiltonian` and `project_state` helpers for applying projection MPOs to DMRG Hamiltonians and initial states.
- Use ITensorMPS `apply` contractions so projected MPO/MPS results keep DMRG-compatible site prime levels.
- Validate that projection MPOs share the target's unprimed site indices before composition.
- Add dense small-system tests for `P' * H * P` masking, projected-state amplitudes, infeasible all-zero projections, and bond-dimension/prime-level expectations.

## Tests

- `git diff --check`
- `julia +1.12 --project=test -e 'push!(LOAD_PATH, pwd()); using Test, Random, LinearAlgebra, TenSolver; include("test/projection_mpo.jl")'`
- `julia +1.12 --project=. -e 'using Pkg; Pkg.test(; coverage=false)'`

Note: the full package test re-resolved into a temporary Julia 1.12 test environment because the checked-in manifest was resolved with Julia 1.10.11.

## Branch Hygiene

- Base branch: `main`
- Source branch point: refreshed `origin/main` after #60 merged via PR #91
- Stacked status: unstacked; this PR targets default `main`
- Prerequisites: #58, #59, and #60 are already integrated into `main`

Closes #61

### CODE REVIEW

**Review by bernalde — COMMENTED** [review]
Note: I'm the PR author, so this account can't submit a formal APPROVE/REQUEST_CHANGES review — posting as COMMENT with the review findings.

## Review

`project_hamiltonian` and `project_state` are correct. I traced the composition logic: `project_hamiltonian` folds each projection `P` in with `H_eff = apply(P, H_eff); H_eff = apply(H_eff, P)`, which — because each `P` produced by `projection_mpo`/`projection_mpos` is a real diagonal 0/1 projector — is equivalent to sandwiching the combined projector `P1*P2*...` on both sides, exactly matching the CoTenN-style `P'HP` construction described in the docstring/comment. `project_state` applies each `P` once, correctly zeroing amplitude on any basis state rejected by any projection.

This is backed by exhaustive dense checks in the new `Projected Hamiltonian and state utilities` and `Infeasible projections remain zero` testsets ([test/projection_mpo.jl](test/projection_mpo.jl)): every bra/ket matrix element of `H_eff` over a 3-site system is checked against brute-force feasibility masking of the exact Hamiltonian, every basis amplitude of the projected state is checked against feasibility, prime-level structure is checked at `plev=0`, `plev=1`, and `plev=2` (confirming no stray primed indices leak from the two `apply` calls), and a fully-infeasible constraint is checked to zero out `H_eff` and the projected state exactly.

### Tests run locally (PR head, clean tree)
- `git diff --check` against merge-base — clean, no whitespace issues.
- `julia --project=. -e 'using Test, Random, LinearAlgebra, TenSolver, ITensors, ITensorMPS; include("test/projection_mpo.jl")'` (Julia 1.10, matching the checked-in manifest) — `1828/1828` pass.
- `julia --project=. -e 'using Pkg; Pkg.test(; coverage=false)'` — full suite passes, including Aqua and doctests.
- GitHub checks for head `948ec20`: all green (macOS/Linux/Windows × Julia lts/1/pre, `build`, `documenter/deploy`).

### Nonblocking: restore docstrings for `project_hamiltonian`/`project_state` in `docs/src/api.md`

The second commit (`fix: keep projection utilities out of public docs`) replaced the docstrings on `project_hamiltonian`/`project_state` with plain `#` comments because the first commit's `Documentation` check failed with `2 docstrings not included in the manual` (I pulled the actual CI log for [the failing run](https://github.com/SECQUOIA/TenSolver.jl/actions/runs/28906012777/job/85753164688) to confirm the exact Documenter error).

That's a real build failure, but the established fix for it elsewhere in this file is to add the function to the `Internal Functions` section of [docs/src/api.md](docs/src/api.md) — that's exactly how `dfa_to_mpo`, `projection_mpo`, `projection_mpos`, and `SparseTensorEntry` are handled: unexported, but documented for "advanced users who may need to understand the internals," explicitly "subject to change without warning." `project_hamiltonian`/`project_state` are the two primary deliverables of this PR (named directly in issue #61's acceptance criteria), not low-level plumbing like `tensor_from_nonzeros`/`dfa_site_tensor` that legitimately have no docstring. Stripping their docs to source comments means `?TenSolver.project_hamiltonian` in the REPL now returns nothing useful, and the generated site loses their usage docs — a real discoverability regression for exactly the two functions future maintainers are most likely to look up when wiring these into `minimize` later (per issue #61's explicit follow-up note).

Suggested fix: restore the two docstrings and add `TenSolver.project_hamiltonian` / `TenSolver.project_state` under a new subsection (e.g. `### Projected Hamiltonian Construction`) in the `Internal Functions` part of `docs/src/api.md`, next to the existing `MPO Construction` subsection.

### Merge readiness
No blocking issues. Issue #61's acceptance criteria are all verifiably met (P'HP masking, zero amplitude on infeasible states, prime-level cleanup, unconstrained tests unchanged, kept unexported/internal), so `Closes #61` is accurate. The one open item is the documentation-consistency suggestion above.

**Comment by bernalde** [comment]
Addressed the review finding `stripped-docstrings-vs-api-md`.

Changes pushed:
- `730a0a0 docs: document projection application helpers`

What changed:
- Restored docstrings for `project_hamiltonian` and `project_state`.
- Added both unexported helpers to the Internal Functions API docs page so Documenter treats them as documented internals.

Verification:
- `git diff --check`
- `julia +1.12 --project=test -e 'push!(LOAD_PATH, pwd()); using Test, Random, LinearAlgebra, TenSolver; include("test/projection_mpo.jl")'`
- `julia +1.12 --project=. -e 'using Pkg; Pkg.test(; coverage=false)'`
- Hosted CI on `730a0a0a8e982e4d84a331c8e5f4d3e2ae39da33`: all test jobs passed; documentation build and `documenter/deploy` passed.

Notes:
- Local `julia +1.12 --project=docs/ docs/make.jl` passed the missing-docs stage, then stopped on a local docs manifest/source-path issue for `MbedTLS_jll`; hosted documentation CI passed on the pushed commit.
- No inline review threads were present, so there were no inline replies to post.

File: src/projection_mpo.jl [hunk 1]
```diff
@@ -261,6 +261,76 @@ end
...
   projection_mpos(Float64, constraints, sites)
 
+"""
+    project_hamiltonian(H, projections; cutoff=1e-8, kwargs...)
+
+Apply one or more diagonal projection MPOs to a Hamiltonian MPO.
+
+For projection MPOs `P`, this builds the CoTenN-style effective Hamiltonian
+`P' * H * P`. The projection MPOs produced by [`projection_mpo`](@ref) are
+real diagonal projectors, so the implementation composes them as `P * H * P`
```
Comment (iagoleal) [inline src/projection_mpo.jl]: We should compose them as `dag(P) * H * P` to make it future-proof. What if we later decide to implement non-diagonal projections? For example, CoTenN's eigenvalue constraints.

File: src/projection_mpo.jl [hunk 2]
```diff
@@ -261,6 +261,76 @@ end
...
+end
+
+"""
+    project_state(psi, projections; cutoff=1e-8, kwargs...)
+
+Apply one or more diagonal projection MPOs to an MPS.
+
+The result has zero amplitude on basis states rejected by any projection, while
+keeping the original unprimed site indices so it can be used as a DMRG input
+state.
```
Comment (iagoleal) [inline src/projection_mpo.jl]:
```suggestion
The result has zero amplitude on basis states rejected by any projection,
while keeping the original unprimed site indices so it can be used as a DMRG input state.
```

File: src/projection_mpo.jl [hunk 3]
```diff
@@ -261,6 +261,76 @@ end
...
+  for P in projection_tuple
+    projected = ITensors.apply(P, projected; cutoff, kwargs...)
+  end
+
+  return projected
+end
+
+projection_sequence(projection::ITensorMPS.MPO) = (projection,)
+projection_sequence(projections::Tuple{Vararg{ITensorMPS.MPO}}) = projections
+projection_sequence(projections::AbstractVector{<:ITensorMPS.MPO}) = Tuple(projections)
```
Comment (iagoleal) [inline src/projection_mpo.jl:311]: Can't we just use an iterator interface to build the tuples?

File: test/projection_mpo.jl [hunk 4]
```diff
@@ -8,6 +8,17 @@ function mpo_diagonal(H, sites, bits)
...
+function mpo_matrix_element(H, sites, bra_bits, ket_bits)
+  bra = ITensorMPS.MPS(sites, string.(bra_bits))
+  ket = ITensorMPS.MPS(sites, string.(ket_bits))
+  return real(ITensors.inner(bra', H, ket))
+end
+
+function mps_amplitude(psi, sites, bits)
+  basis = ITensorMPS.MPS(sites, string.(bits))
+  return real(ITensors.inner(basis, psi))
+end
```
Comment (iagoleal) [inline test/projection_mpo.jl]: These are very useful methods. We should move them to `test/utils.jl` for discoverability.

File: src/projection_mpo.jl [hunk 5]
```diff
@@ -261,6 +261,76 @@ end
...
+"""
+    project_hamiltonian(H, projections; cutoff=1e-8, kwargs...)
+
+Apply one or more diagonal projection MPOs to a Hamiltonian MPO.
+
+For projection MPOs `P`, this builds the CoTenN-style effective Hamiltonian
+`P' * H * P`. The projection MPOs produced by [`projection_mpo`](@ref) are
+real diagonal projectors, so the implementation composes them as `P * H * P`
+using ITensorMPS' `apply` helper to contract the MPOs and restore the usual
+unprimed/primed site-index structure expected by DMRG. The resulting bond
```
Comment (iagoleal) [inline src/projection_mpo.jl]: This phrase is an implementation detail and should not be in the docstring. Docstrings should focus on "what" and "why", not on "how" we are doing it right now.

**Review by iagoleal — CHANGES_REQUESTED** [review]

**Review by iagoleal — APPROVED** [review]
